Feature/flatmesh support

nm2107 opened the pull request:
From the second commit message :

--------

Add support for the `Unwrap Mesh` and `Unwrap Face` features in the Mesh
Workbench.

These features allow to unfold a mesh, as detailled here :
    - https://wiki.freecadweb.org/Mesh_Workbench
    - https://www.youtube.com/watch?v=-KjMZkLfE1A

These features are enabled via the `BUILD_FLAT_MESH` compilation option
(which defaults to `OFF`), and makes use of the `pybind11` library.

This commit also adds the `pybind11` module to the flatpak app.

However, the `FREECAD_USE_PYBIND11` compilation option of FreeCAD enables
some code paths which are requiring the `Python.h` header, but this
header is not available in the flatpak runtime. Due to this reason, the
compilation fails.
Such header is generally found in `python-dev` like packages
(`python-dev` on debian based systems, `python-devel` on fedora based
systems), or directly in the python language installation from sources.

Providing this header to the runtime is necessary, but the python
language installation might not be done in this flatpak application
declaration, but rather in a flatpak extention (e.g. like the
`org.freedesktop.Sdk.Extension.php74` extention which brings the PHP
language installation to the runtime which uses it).


-------


I sent an email to `admins [at] flathub [dot] org` to ask about such python extension, let me know if I should make this request somewhere else.

nm2107 committed:
fix `boost` module URL

As the original one was responding with HTTP 502 status.

nm2107 committed:
add support for flatmesh feature

Add support for the `Unwrap Mesh` and `Unwrap Face` features in the Mesh
Workbench.

These features allow to unfold a mesh, as detailled here :
    - https://wiki.freecadweb.org/Mesh_Workbench
    - https://www.youtube.com/watch?v=-KjMZkLfE1A

These features are enabled via the `BUILD_FLAT_MESH` compilation option
(which defaults to `OFF`), and makes use of the `pybind11` library.

This commit also adds the `pybind11` module to the flatpak app.

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -181,6 +181,8 @@ modules:
         url: https://dl.bintray.com/boostorg/release/1.67.0/source/boost_1_67_0.tar.bz2
         sha256: 2684c972994ee57fc5632e03bf044746f6eb45d4920c343937a465fd67a5adba
+        mirror-urls:
+          - https://boostorg.jfrog.io/artifactory/main/release/1.67.0/source/boost_1_67_0.tar.bz2
```
not sure how the builds have gone recently, but bintray is definitely dead. Please just replace the URL. No need for mirrors.

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -610,6 +612,21 @@ modules:
+      - -DPYBIND11_TEST=OFF
+    sources:
+      - type: archive
+        url: http://deb.debian.org/debian/pool/main/p/pybind11/pybind11_2.9.0.orig.tar.gz
```
I'd rather the archives from pypi.

https://pypi.org/project/pybind11/#files

Also it seems it exist was a platform independent wheel.

nm2107 replied:
Thank you for the review.
I tried to install the dep with pypi, but then the FreeCAD compilation step is failing with this error message : 

```
CMake Error at cMake/FreeCAD_Helpers/SetupPybind11.cmake:7 (find_package):
  By not providing "Findpybind11.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "pybind11",
  but CMake did not find one.

  Could not find a package configuration file provided by "pybind11" with any
  of the following names:

    pybind11Config.cmake
    pybind11-config.cmake

  Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
  "pybind11_DIR" to a directory containing one of the above files.  If
  "pybind11" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:47 (SetupPybind11)
```

So I'd rather keep this way to install this dep.


For the record, here was how the dep install via pypi was done : 

```yaml
  # pybind11 used by mesh flattening feature
  - name: pybind11
    buildsystem: simple
    build-commands:
      - pip3 install --exists-action=i --no-index --no-build-isolation --find-links="file://${PWD}" --prefix=${FLATPAK_DEST} pybind11
    sources:
      - type: file
        url: https://files.pythonhosted.org/packages/55/08/20f915ceea974f9e6a4aa61b515dac84befda136c4d03a873550f19540cf/pybind11-2.9.0.tar.gz
        sha256: 46d315a669a59b660def6c5ca4e128f6fa975ec4abce6019a43d92084ad1bf89
```

hfiguiere replied:
alternatively does the archive from pypi work when using cmake? I would expect it to be the same

nm2107 replied:
Not at all, the archive is not the same. I first changed the archive url and tried to install it, but it is not meant to be used with cmake. It is only an archive meant to be installed via pip.

hfiguiere commented:
In the **Sdk** the Python.h is in `/usr/include/python3.9/Python.h`. Which is the usual place. Not sure I understand what's the problem.

nm2107 commented:
> In the **Sdk** the Python.h is in `/usr/include/python3.9/Python.h`. Which is the usual place. Not sure I understand what's the problem.

Mh, indeed, the file exists, thank you !

I'm trying to set FreeCAD compilation options to use this right now :)

hfiguiere commented:
note that the current build failure is cause by a bug in flatpak-builder that was fixed and deployed, and I'm waiting for confirmation of other builds to retrigger.

hfiguiere commented:
> > In the **Sdk** the Python.h is in `/usr/include/python3.9/Python.h`. Which is the usual place. Not sure I understand what's the problem.
> 
> Mh, indeed, the file exists, thank you !
> 
> I'm trying to set FreeCAD compilation options to use this right now :)

Given all the python we build for FreeCAD, I'm wondering what the problems are here.

nm2107 commented:
> Given all the python we build for FreeCAD, I'm wondering what the problems are here.


Note that the issue appear even when setting only the `-DFREECAD_USE_PYBIND11=ON` build option :thinking:  .
Probably some conditionals in the codebase relying on that.

nm2107 commented:
I managed to compile FreeCAD by indicating where to look for the `Python.h` header.
In the flatpak SDK, I ran :

```bash
$ pkg-config --cflags python3
-I/usr/include/x86_64-linux-gnu/python3.9 -I/usr/include/python3.9
```

and I added these paths to both C++ and C compilers flags via cmake options (as both compilers are used to compile FreeCAD).


The features are now accessible :) . Thank you for your help !
I updated the PR.

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -627,10 +644,14 @@ modules:
             - -DSHIBOKEN_LIBRARY=shiboken2.cpython-39-x86_64-linux-gnu
             - -DPYSIDE_LIBRARY=pyside2.cpython-39-x86_64-linux-gnu
+            - -DCMAKE_C_FLAGS_INIT=-I/usr/include/x86_64-linux-gnu/python3.9 -I/usr/include/python3.9
+            - -DCMAKE_CXX_FLAGS_INIT=-I/usr/include/x86_64-linux-gnu/python3.9 -I/usr/include/python3.9
```
ideally there should be a patch for upstream FreeCAD to actually call Python pkg-config and use that.
(I'm saying this for the future, it's good here for now)

nm2107 replied:
I was digging in the FreeCAD source and also in this flatpak recipe, and noticed that the `pybind11` dependency has the same purpose than the `boost` dependency (i.e. python bidings for c++).

So I then tried to not include the `pybind11` module in the flatpak recipe, and only specify the `-DBUILD_FLAT_MESH=ON` flag on FreeCAD module build, as FreeCAD has conditionals in its source code to fallback to `boost` when `pybind11` is not used (i.e. when `FREECAD_USE_PYBIN11` option is let to `OFF`), but I've got this compilation error when building FreeCAD : 

```
FAILED: src/App/CMakeFiles/FreeCADApp.dir/DocumentObserverPython.cpp.o 
/usr/bin/ccache /usr/bin/c++ -DDOCDIR=\"/app/share/doc/FreeCAD\" -DFreeCADApp_EXPORTS -DHAVE_CONFIG_H -DQT_CORE_LIB -DQT_NO_DEBUG -DQT_XML_LIB -DRESOURCEDIR=\"/app/share\" -D_OCC64 -I/run/build/FreeCAD/_flatpak_build -I/run/build/FreeCAD/_flatpak_build/src -I/run/build/FreeCAD/src -I/run/build/FreeCAD/_flatpak_build/src/App -I/usr/include/python3.9 -isystem /usr/include/QtCore -isystem /usr/lib/mkspecs/linux-g++ -isystem /usr/include/QtXml -Wall -Wextra -Wno-write-strings -O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -O2 -g -DNDEBUG -fPIC -pthread -fPIC -std=gnu++11 -MD -MT src/App/CMakeFiles/FreeCADApp.dir/DocumentObserverPython.cpp.o -MF src/App/CMakeFiles/FreeCADApp.dir/DocumentObserverPython.cpp.o.d -o src/App/CMakeFiles/FreeCADApp.dir/DocumentObserverPython.cpp.o -c /run/build/FreeCAD/src/App/DocumentObserverPython.cpp
In file included from /usr/include/c++/11.2.0/iterator:65,
                 from /app/include/boost/iterator/iterator_categories.hpp:23,
                 from /app/include/boost/iterator/iterator_adaptor.hpp:12,
                 from /app/include/boost/iterator/reverse_iterator.hpp:10,
                 from /app/include/boost/signals2/detail/auto_buffer.hpp:21,
                 from /app/include/boost/signals2/connection.hpp:21,
                 from /app/include/boost/signals2/signal.hpp:22,
                 from /app/include/boost/signals2.hpp:19,
                 from /run/build/FreeCAD/src/boost_signals2.hpp:6,
                 from /run/build/FreeCAD/src/App/Application.h:27,
                 from /run/build/FreeCAD/src/App/DocumentObserverPython.cpp:30:
/usr/include/c++/11.2.0/bits/stream_iterator.h:68:49: internal compiler error: Segmentation fault
   68 |       : _M_stream(0), _M_value(), _M_ok(false) {}
      |                                                 ^
0x1886451 diagnostic_impl(rich_location*, diagnostic_metadata const*, int, char const*, __va_list_tag (*) [1], diagnostic_t)
	???:0
0x1886826 internal_error(char const*, ...)
	???:0
0xda002e crash_signal(int)
	???:0
0x921adb ggc_set_mark(void const*)
	???:0
0xb419a4 gt_ggc_mx_gimple_df(void*)
	???:0
0xb41bb3 gt_ggc_mx_function(void*)
	???:0
0x882e70 gt_ggc_mx_lang_tree_node(void*)
	???:0
0x881faa gt_ggc_mx_lang_tree_node(void*)
	???:0
0x881faa gt_ggc_mx_lang_tree_node(void*)
	???:0
0x881faa gt_ggc_mx_lang_tree_node(void*)
	???:0
0x881faa gt_ggc_mx_lang_tree_node(void*)
	???:0
0xb3a8e1 gt_ggc_mx_tree_statement_list_node(void*)
	???:0
0x881f45 gt_ggc_mx_lang_tree_node(void*)
	???:0
0x881faa gt_ggc_mx_lang_tree_node(void*)
	???:0
0x882eca gt_ggc_mx_lang_tree_node(void*)
	???:0
0xb39c81 gt_ggc_mx_vec_tree_va_gc_(void*)
	???:0
0x8839b8 gt_ggc_mx_lang_type(void*)
	???:0
0x882bd5 gt_ggc_mx_lang_tree_node(void*)
	???:0
0x8828f5 gt_ggc_mx_lang_tree_node(void*)
	???:0
0x882bc3 gt_ggc_mx_lang_tree_node(void*)
	???:0
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
```

So I guess the `BUILD_FLAT_MESH` option really requires `pybind11`.


Alternatively, I also tried to only use `pybind11` (i.e. omit `boost` module in the flatpak recipe), but again, there is a build error :

```
========================================================================
Building module coin in ./.flatpak-builder/build/coin-1
========================================================================
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Boost (missing: Boost_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.22/Modules/FindBoost.cmake:2360 (find_package_handle_standard_args)
  CMakeLists.txt:160 (find_package)


-- Configuring incomplete, errors occurred!
See also "/run/build/coin/_flatpak_build/CMakeFiles/CMakeOutput.log".
Error: module coin: Child process exited with code 1
```

So it seems that there is a need for both `boost` and `pybind11` libraries.

nm2107 replied:
Actually, I figured out that I have a misfunctionning RAM chip that I just added on my computer, hence the segfault :/

I'll try again w/o the RAM chip to see how it goes.

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -610,6 +610,21 @@ modules:
+      - -DPYBIND11_TEST=OFF
+    sources:
+      - type: archive
+        url: http://deb.debian.org/debian/pool/main/p/pybind11/pybind11_2.9.0.orig.tar.gz
```
```suggestion
        url: https://deb.debian.org/debian/pool/main/p/pybind11/pybind11_2.9.0.orig.tar.gz
```

at least let's get it over https.

nm2107 replied:
done ;)

hfiguiere commented:
How do I test this? (I'm still somewhat newbie in FreeCAD)

Just want to make sure all is OK

nm2107 commented:
You can reproduce what's presented in the video linked in the PR description in order to test this feature :) (i.e. create a body with concurrent faces, select them, create a shapebinder from the selected faces, then turn it into a mesh and flatten the mesh).

I'll keep you informed whether or not the build with `boost` only is successful ;)

nm2107 committed:
avoid to use pybind11 library

As there is already the `boost` library in use, and both serve the same
aim (i.e. python bidings for C++).

The flat mesh feature can rely on the `boost` library in order to work.

nm2107 commented:
The build without `pybind11` was working, and I'm able to use the flatmesh feature too :)

Looks like this PR was very small actually ^^. I added a commit to remove the use of `pybind11`.

Thank you for your help, feel free to merge as you'd like to :)

hfiguiere commented:
indeed.

hfiguiere commented:
bot, build

hfiguiere commented:
merged. Thanks a lot !
